File: server/taskwright_server/service/hub.py
"""事件分发：每个任务一个。管 SSE 订阅者、库事件的转发、兜底轮询与 Last-Event-ID 补发（约定第 2、3 节）。

库事件的正本是库里的事件表。后端记着「已经转发到第几号」，三种提示会让它去查库：写入类工具的
tool_execution_end、扩展命令经状态栏带回的结果、pi 在跑期间每 2 秒一次的兜底轮询。查到新行就按序号现拼内容推送。
后端重启、pi 重启、断线补发走同一条路：从第 N 号之后查事件表、现拼内容，不依赖内存里存过什么。
"""

from __future__ import annotations

import queue
import sqlite3
import threading
import time
from pathlib import Path

from taskwright_observatory import taskdb
from taskwright_server.service import library

REPLAY_WINDOW = 500
POLL_SECONDS = 2.0


class Subscriber:
    def __init__(self, session: str | None):
        self.session = session
        self.queue: queue.Queue = queue.Queue()
        self.last_seq = 0


class Hub:
    def __init__(self, task_dir: Path, is_running=lambda: False):
        self.task_dir = Path(task_dir)
        self.is_running = is_running
        self.lock = threading.Lock()
        self.subscribers: list[Subscriber] = []
        self.forwarded = self.current_max()
        self.stats = {"轮询次数": 0, "轮询发现新行次数": 0, "提示次数": 0, "推送的库事件数": 0}
        self._stop = threading.Event()
        threading.Thread(target=self._poll_loop, daemon=True).start()

    def current_max(self) -> int:
        """SELECT MAX(seq)：每次新开连接、查完就关，不留读事务（实测长开读事务会让 -wal 文件一直变大）。"""
        path = library.db_file(self.task_dir)
        if not path.is_file():
            return 0
        conn = taskdb.open_readonly(path)
        try:
            return int(conn.execute("SELECT COALESCE(MAX(seq), 0) FROM event").fetchone()[0])
        except sqlite3.Error:
            return 0
        finally:
            conn.close()

    # ───────────── 推送 ─────────────

    def emit(self, name: str, data: dict) -> None:
        """过程与对话类事件：没有 id，漏了不补。订阅时带了 session 的只收自己会话的。"""
        for sub in list(self.subscribers):
            if sub.session and data.get("session_id") and data["session_id"] != sub.session:
                continue
            sub.queue.put((name, None, data))

    def trigger(self) -> None:
        """有提示说库里可能有新行：查、拼、推，更新已转发序号。"""
        self.stats["提示次数"] += 1
        with self.lock:
            events, top = library.library_events(self.task_dir, self.forwarded)
            for name, data in events:
                for sub in list(self.subscribers):
                    sub.queue.put((name, data["seq"], data))
                self.stats["推送的库事件数"] += 1
            self.forwarded = max(self.forwarded, top)

    def _poll_loop(self) -> None:
        while not self._stop.wait(POLL_SECONDS):
            if not self.is_running():
                continue
            self.stats["轮询次数"] += 1
            if self.current_max() > self.forwarded:
                self.stats["轮询发现新行次数"] += 1
                self.trigger()

    # ───────────── 订阅 ─────────────

    def subscribe(self, session: str | None, last_event_id: int | None) -> tuple[Subscriber, list]:
        """先登记订阅者，再算补发：带了 Last-Event-ID 的，从它之后补；差距超过 500 条发 resync。"""
        sub = Subscriber(session)
        with self.lock:
            self.subscribers.append(sub)
            replay: list = []
            if last_event_id is not None:
                top = self.current_max()
                if top - last_event_id > REPLAY_WINDOW:
                    replay = [("resync", None, {"reason": "gap_too_large"})]
                elif top > last_event_id:
                    events, _ = library.library_events(self.task_dir, last_event_id)
                    replay = [(name, data["seq"], data) for name, data in events]
                sub.last_seq = last_event_id
        return sub, replay

    def unsubscribe(self, sub: Subscriber) -> None:
        with self.lock:
            if sub in self.subscribers:
                self.subscribers.remove(sub)

    def close(self) -> None:
        self._stop.set()

File: backend/src/hub.ts
/**
 * 事件分发：每个任务一个。管 SSE 订阅者、库事件的转发、兜底轮询与 Last-Event-ID 补发。
 *
 * 库事件的正本是库里的事件表。后端记着「已经转发到第几号」，三种提示会让它去查库：写入类工具的 tool_execution_end、
 * 扩展命令经状态栏带回的结果、pi 在跑期间每 2 秒一次的兜底轮询。查到新行就按序号现拼内容推送。
 * 后端重启、pi 重启、断线补发走同一条路：从第 N 号之后查事件表、现拼内容，不依赖内存里存过什么。
 */

import * as library from "./library.ts";

export const hubSettings = {
  /** 补发窗口：断线重连时差得比这更多，就不补发，改发一条 resync。 */
  replayWindow: 500,
  /** 兜底轮询的间隔（毫秒）。 */
  pollMs: 2000,
};

export type HubItem = [string, number | null, Record<string, any>];

/** 一个订阅者：它自己的事件队列，以及已经发给它的最大库事件序号。 */
export class Subscriber {
  readonly session: string | null;
  lastSeq = 0;
  private items: HubItem[] = [];
  private waiter: ((item: HubItem | null) => void) | null = null;

  constructor(session: string | null) {
    this.session = session;
  }

  put(item: HubItem): void {
    if (this.waiter) {
      const waiter = this.waiter;
      this.waiter = null;
      waiter(item);
    } else {
      this.items.push(item);
    }
  }

  /** 取下一条；超时（毫秒）时为 null。 */
  get(timeoutMs: number): Promise<HubItem | null> {
    if (this.items.length) return Promise.resolve(this.items.shift()!);
    return new Promise((done) => {
      const timer = setTimeout(() => {
        this.waiter = null;
        done(null);
      }, timeoutMs);
      this.waiter = (item) => {
        clearTimeout(timer);
        done(item);
      };
    });
  }

  /** 连接断了：叫醒正在等的读取，让它退出。 */
  wake(): void {
    if (this.waiter) {
      const waiter = this.waiter;
      this.waiter = null;
      waiter(null);
    }
  }
}

export class Hub {
  readonly taskDir: string;
  isRunning: () => boolean;
  readonly subscribers: Subscriber[] = [];
  forwarded: number;
  readonly stats = { 轮询次数: 0, 轮询发现新行次数: 0, 提示次数: 0, 推送的库事件数: 0 };
  private timer: NodeJS.Timeout;
  /** 服务已经收尾：事件流不再等新事件。 */
  closed = false;

  constructor(taskDir: string, isRunning: () => boolean = () => false) {
    this.taskDir = taskDir;
    this.isRunning = isRunning;
    this.forwarded = library.currentMax(taskDir);
    this.timer = setInterval(() => this.poll(), hubSettings.pollMs);
    this.timer.unref();
  }

  /** 过程与对话类事件：没有 id，漏了不补。订阅时带了 session 的只收自己会话的。 */
  emit(name: string, data: Record<string, any>): void {
    for (const sub of [...this.subscribers]) {
      if (sub.session && data.session_id && data.session_id !== sub.session) continue;
      sub.put([name, null, data]);
    }
  }

  /** 有提示说库里可能有新行：查、拼、推，更新已转发序号。 */
  trigger(): void {
    this.stats.提示次数 += 1;
    const [events, top] = library.libraryEvents(this.taskDir, this.forwarded);
    for (const [name, data] of events) {
      for (const sub of [...this.subscribers]) sub.put([name, data.seq, data]);
      this.stats.推送的库事件数 += 1;
    }
    this.forwarded = Math.max(this.forwarded, top);
  }

  private poll(): void {
    if (!this.isRunning()) return;
    this.stats.轮询次数 += 1;
    if (library.currentMax(this.taskDir) > this.forwarded) {
      this.stats.轮询发现新行次数 += 1;
      this.trigger();
    }
  }

  /** 先登记订阅者，再算补发：带了 Last-Event-ID 的，从它之后补；差距超过补发窗口发 resync。 */
  subscribe(session: string | null, lastEventId: number | null): [Subscriber, HubItem[]] {
    const sub = new Subscriber(session);
    this.subscribers.push(sub);
    let replay: HubItem[] = [];
    if (lastEventId !== null) {
      const top = library.currentMax(this.taskDir);
      if (top - lastEventId > hubSettings.replayWindow) {
        replay = [["resync", null, { reason: "gap_too_large" }]];
      } else if (top > lastEventId) {
        const [events] = library.libraryEvents(this.taskDir, lastEventId);
        replay = events.map(([name, data]) => [name, data.seq, data]);
      }
      sub.lastSeq = lastEventId;
    }
    return [sub, replay];
  }

  unsubscribe(sub: Subscriber): void {
    const i = this.subscribers.indexOf(sub);
    if (i >= 0) this.subscribers.splice(i, 1);
    sub.wake();
  }

  /** 服务收尾的第一步：停掉兜底轮询；事件流照旧开着，之后关 pi 时推的执行者状态还送得到。 */
  stopPolling(): void {
    clearInterval(this.timer);
  }

  /** 服务收尾的最后一步：停掉轮询，叫醒各条事件流让它们写完手上的事件、正常结束连接。 */
  close(): void {
    clearInterval(this.timer);
    this.closed = true;
    for (const sub of [...this.subscribers]) sub.wake();
  }
}
